### PR TITLE
Refactor baselines to allow for extension

### PR DESCRIPTION
**Pull request by erikwijmans**
## Motivation and Context

The current PPO baselines code isn't easily extendible to other tasks, architectures, or variants of on-policy RL.  This PR does some refactoring to allow other architectures and variants of on-policy RL to be added and helps slightly with other tasks.

This PR also introduces support for N-layer GRUs and support for N-layer LSTMs as the state-encoder whereas the current code is limited to a single layer GRU.

This PR also adds support for making the state-encoder auto-regressive (i.e. taking the previous action as input).

I also made this script to remap any current checkpoints so they can be loaded: https://gist.github.com/erikwijmans/94f33461982ac9e3cc354fb26e4b80e1

## How Has This Been Tested

Via the tests and running `train_ppo.py`, and `evaluate_ppo.py`.  I also

## Types of changes

- Breaking change (fix or feature that would cause existing functionality to change)

### CODE REVIEW

**Comment by kowshikthopalli** [comment]
I am not sure I can post this. This is great Erik. Simplifies a lot of complex structure in the present code and introduces allowance for  extension to desired models.
I would request you to consider adding multi-gpu support for models and simulator in this code.
I am currently using  ideas from this closed issue -https://github.com/facebookresearch/habitat-api/issues/38#issuecomment-480515073 for multiple gpus on simulator and I think it will be nice to add in this refactored code.

**Comment by erikwijmans** [comment]
I have something else I've cooked up for nice clean multi-gpu support.  Stay tuned :)

**Comment by erikwijmans** [comment]
> I am not sure I can post this.

You are certainly welcome to post things!  The more community involvement the better!

File: habitat_baselines/rl/models/simple_cnn.py [hunk 1]
```diff
@@ -0,0 +1,138 @@
...
+                    stride=self._cnn_layers_stride[1],
+                ),
+                nn.ReLU(True),
+                nn.Conv2d(
+                    in_channels=64,
+                    out_channels=32,
+                    kernel_size=self._cnn_layers_kernel_size[2],
+                    stride=self._cnn_layers_stride[2],
+                ),
+                #  nn.ReLU(True),
```
Comment (erikwijmans) [inline habitat_baselines/rl/models/simple_cnn.py]: @abhiskk -- When refactoring the baseline code, I notice we are missing this ReLU.  There is non-linearity between the last conv and the FC layer!

File: habitat_baselines/rl/models/rnn_state_encoder.py [hunk 2]
```diff
@@ -0,0 +1,116 @@
+import torch
+import torch.nn as nn
+
+
+class RNNStateEncoder(nn.Module):
+    def __init__(self, input_size, hidden_size, num_layers=1, rnn_type="GRU"):
```
Comment (mathfac) [inline habitat_baselines/rl/models/rnn_state_encoder.py]: Maybe we can add doc string here and mention what are options for `rnn_type`  and add typing.

File: habitat_baselines/rl/models/rnn_state_encoder.py [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+        hidden_states = self._unpack_hidden(hidden_states)
+        x, hidden_states = self.rnn(
+            x.unsqueeze(0),
+            self._mask_hidden(hidden_states, masks.unsqueeze(0)),
+        )
+        x = x.squeeze(0)
+        hidden_states = self._pack_hidden(hidden_states)
+        return x, hidden_states
+
+    def seq_forward(self, x, hidden_states, masks):
```
Comment (mathfac) [inline habitat_baselines/rl/models/rnn_state_encoder.py]: Definitely doc string with explanation needed.

File: habitat_baselines/rl/models/simple_cnn.py [hunk 4]
```diff
@@ -0,0 +1,138 @@
+import numpy as np
```
Comment (mathfac) [inline habitat_baselines/rl/models/simple_cnn.py]: Maybe all baselines can share model architectures and move it to `habitat_baselines/models`.

File: habitat_baselines/rl/models/simple_cnn.py [hunk 5]
```diff
@@ -0,0 +1,138 @@
...
+        else:
+            self._n_input_rgb = 0
+
+        if "depth" in observation_space.spaces:
+            self._n_input_depth = observation_space.spaces["depth"].shape[2]
+        else:
+            self._n_input_depth = 0
+
+        # kernel size for different CNN layers
+        self._cnn_layers_kernel_size = [(8, 8), (4, 4), (3, 3)]
```
Comment (mathfac) [inline habitat_baselines/rl/models/simple_cnn.py]: Maybe worth to make method arguments with default values.

File: habitat_baselines/rl/models/simple_cnn.py [hunk 6]
```diff
@@ -0,0 +1,138 @@
+import numpy as np
+import torch
+import torch.nn as nn
+
+from habitat_baselines.rl.ppo.utils import Flatten
+
+
+class SimpleCNN(nn.Module):
```
Comment (mathfac) [inline habitat_baselines/rl/models/simple_cnn.py]: Doc string needed, it is SimpleCNN adopted for Habitat actually. Maybe worth to name it `SimpleCNNPerception`.

File: habitat_baselines/train_ppo.py [hunk 7]
```diff
@@ -157,13 +165,127 @@ def construct_envs(args):
...
+
+    value_loss, action_loss, dist_entropy = agent.update(rollouts)
+
+    rollouts.after_update()
+    return time() - t_update_model, value_loss, action_loss, dist_entropy
+
+
 def run_training():
     parser = ppo_args()
+    parser.add_argument(
```
Comment (mathfac) [inline habitat_baselines/train_ppo.py]: Why not to add to `ppo_args` ?

File: habitat_baselines/rl/models/simple_cnn.py (hunk 5 above)
Comment (erikwijmans) [reply habitat_baselines/rl/models/simple_cnn.py]: The logic is hard coded to the length of this list.

File: habitat_baselines/train_ppo.py (hunk 7 above)
Comment (erikwijmans) [reply habitat_baselines/train_ppo.py]: The `sim-gpu-id` and `pth-gpu-id` are specific to this train script.  Different algos may have different ways to figure out which GPU to use (i.e. DD-PPO).

File: habitat_baselines/rl/models/simple_cnn.py (hunk 4 above)
Comment (erikwijmans) [reply habitat_baselines/rl/models/simple_cnn.py]: Not all baselines use DL tho

**Comment by mathfac** [comment]
@erikwijmans rebase needed based on PR https://github.com/facebookresearch/habitat-api/pull/153.

**Comment by erikwijmans** [comment]
It already is :)

**Comment by kowshikthopalli** [comment]
> I have something else I've cooked up for nice clean multi-gpu support. Stay tuned :)

Hi Erik. Was wondering if you have updated the code with multi-gpu support for models somewhere.

**Pull request merged**